### PR TITLE
Populate conditionally rendered fields

### PR DESCRIPTION
**Pull request by qroll**
**Changes**

- Track user's historical field inputs in a `FormValues` context
- Update fields on changes (via `watch`) and reset
- Introduce a `restoreMode` that configures how conditionally rendered fields are populated when they are toggled
- Added and reorganised stories under `Condtional Rendering/Rules` and `Condtional Rendering/Restoring Values`
- [delete] branch

### CODE REVIEW

File: src/components/frontend-engine/types.ts [hunk 1]
```diff
@@ -106,6 +107,7 @@ export type TFrontendEngineFieldSchema<V = undefined> =
 export type TFrontendEngineValues<T = any> = Record<keyof T, T[keyof T]>;
 export type TRevalidationMode = Exclude<keyof ValidationMode, "onTouched" | "all">;
 export type TValidationMode = keyof ValidationMode;
+export type TRestoreMode = "default-value" | "user-input";
```
Comment (weili-govtech) [inline src/components/frontend-engine/types.ts:110]: apologies, I missed this out, can we have another mode to not restore the values, this is to support our current behaviour of no restoring the values at all

File: src/components/frontend-engine/types.ts [hunk 2]
```diff
@@ -74,6 +74,7 @@ export interface IFrontendEngineData<V = undefined> {
 	validationMode?: TValidationMode | undefined;
 	/** Additional properties to mutate the sections schema on-the-fly */
 	overrides?: Record<string, RecursivePartial<TFrontendEngineFieldSchema<V>>> | undefined;
+	restoreMode?: TRestoreMode | undefined;
```
Comment (weili-govtech) [inline src/components/frontend-engine/types.ts:77]: can add jsdocs comment to explain what is `restoreMode`

File: src/components/frontend-engine/form-values/context-provider.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { Dispatch, MutableRefObject, ReactElement, SetStateAction, createContext, useRef, useState } from "react";
+
```
Comment (weili-govtech) [inline src/components/frontend-engine/form-values/context-provider.tsx:2]: add simple explanation about the need to use form values context

File: src/components/fields/chips/chips.tsx [hunk 4]
```diff
@@ -27,12 +28,19 @@ export const Chips = (props: IGenericFieldProps<IChipsSchema>) => {
...
+	const { setValue, unregister } = useFormContext();
 	const { setFieldValidationConfig, removeFieldValidationConfig } = useValidationConfig();
 
 	// =============================================================================
 	// EFFECTS
 	// =============================================================================
+	useEffect(() => {
+		if (!showTextarea) {
+			// at the start, the textarea's default value has to be manually cleared from the form state
+			unregister(getTextareaId());
```
Comment (weili-govtech) [inline src/components/fields/chips/chips.tsx:40]: just to clarify, is this a temp fix before you make a complete fix in another PR?
I wonder if `unregister` can be used in FieldWrapper or ConditionalRenderer

**Review by weili-govtech — CHANGES_REQUESTED** [review]
do pull the latest commits from master because there is an update to the filter clear button that is essentially resetting the fields under it

File: src/components/fields/chips/chips.tsx (hunk 4 above)
Comment (qroll) [reply src/components/fields/chips/chips.tsx:40]: yup this is temporary. I'm intending to run `unregister` for each `idsToDelete` in `ConditionalRenderer`. `FieldWrapper` won't work for all cases as it requires the component to unmount which doesn't happen on fields nested within conditionally rendered fields

File: src/components/frontend-engine/types.ts (hunk 1 above)
Comment (qroll) [reply src/components/frontend-engine/types.ts]: added a `"none"` mode which clears the values similarly to the reset button

**Review by weili-govtech — APPROVED** [review]
